### ISSUE TITLE
StandardScaler on returns ? Is this correct ?

### ISSUE THREAD

**Issue by makovez** [kieranjwood/trading-momentum-transformer#5]
Please accept my apologies for the frequent messages. I am writing to you regarding a doubt that has arisen after analyzing your code.

While reviewing your code, I observed that you have used StandardScaler on both features and the return target variable in the function `set_scalers(self, df)`. This approach is perfectly acceptable. However, I have concerns regarding the calculation of captured returns in the custom loss function defined in `SharpeLoss` class inside deep_momentum_network file.

Specifically, the function `call(self, y_true, weights)` uses scaled returns in the first argument. The captured returns are then calculated using this scaled return and the position size, `captured_returns = weights * y_true`. This results in a return that is not accurate since it is based on scaled returns rather than actual returns.

Furthermore, there is no provision to unscale the returns or position in the code as the only function that does that is `format_predictions(self, predictions)`, which does never get called.

**Comment by MickyDowns** [comment]
Hey @makovez, take a look at how that SharpeLoss.call output is being used. It's not used to report the return of predictions on the test set. Nor is it used to evaluate performance on the validation set. It's being used to determine if the lowest level model fit is improving. For that, all that it needs to "know" is that losses are headed in the right direction and, if so, to continue fitting.

That is why that SharpeLoss calculation is distinct from the SharpeValidationLoss calculation in deep_momentum_network which uses positions * actual returns and for the performance reporting calculations in get_positions() which uses position predictions * actual returns.

Note: when you dig into the SharpeValidationLoss function you'll likely encounter volatility re-scaling calculations used to determine positions. These are used to ensure the super-model is selecting sub-models that appropriately factor-in each asset's risk (volatility). It avoids selecting models merely b/c they have the greatest return in the lookback window.

**Comment by makovez** [comment]
Hi, 

Thanks for your answer. 

> It's not used to report the return of predictions on the test set. 
I never said this. I said the issue is with scaling of returns.

Returns is the target variable, and it's being scaled, before everything else is done. So whether you are looking at  SharpeValidationLoss, get_positions(), SharpeLoss.call every time you see the word "returns" that's the scaled one and not the original real return. Which obviously would lead to a different captured return. 

Wherever you are in the code, data is taken from this instance of this class 
![image](https://user-images.githubusercontent.com/21694707/226132577-dc14f0e7-35dc-4f94-9728-2f72f29b6d60.png)

For example with either: `model_features.test_sliding, model_features.test_fixed, model_features.train, model_features.valid`

These are being defined here below
![image](https://user-images.githubusercontent.com/21694707/226132677-2a50dc9a-a155-4ff1-b262-3cfb14752af8.png)

As you can see in this snippet of the code: 

```
train, valid, test, test_with_buffer = [
            self.transform_inputs(data)
            for data in [train, valid, test, test_with_buffer]
        ] 
```

For each data, every time, self.transform_inputs(data) is called, which in addition to scaling all the features, also scales the target variable.

**Comment by makovez** [comment]
Thanks for your answer. 

Actually I just notice that TARGET is never being scaled. Because it's always being excluded from scaling.

**Comment by MickyDowns** [comment]
target_returns are scaled, just not with standardScaler in model_inputs.modelFeatures(). It's done in data_prep via classical_strategies.calc_vol_scaled_returns() for the reasons above.
